packages/aws-cdk/lib/api/hotswap/code-build-projects.ts

```diff
@@ -1,62 +1,81 @@
 import * as AWS from 'aws-sdk';
 import { ISDK } from '../aws-auth';
 import { EvaluateCloudFormationTemplate } from '../evaluate-cloudformation-template';
-import { ChangeHotswapImpact, ChangeHotswapResult, HotswapOperation, HotswappableChangeCandidate, lowerCaseFirstCharacter, transformObjectKeys } from './common';
+import { ChangeHotswapResult, classifyChanges, HotswappableChangeCandidate, lowerCaseFirstCharacter, transformObjectKeys } from './common';
 
 export async function isHotswappableCodeBuildProjectChange(
   logicalId: string, change: HotswappableChangeCandidate, evaluateCfnTemplate: EvaluateCloudFormationTemplate,
 ): Promise<ChangeHotswapResult> {
   if (change.newValue.Type !== 'AWS::CodeBuild::Project') {
-    return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+    return [];
+  }
+
+  const ret: ChangeHotswapResult = [];
+
+  const { hotswappableProps, nonHotswappableProps } = classifyChanges(change, ['Source', 'Environment', 'SourceVersion']);
+
+  const nonHotswappablePropNames = Object.keys(nonHotswappableProps);
+  if (nonHotswappablePropNames.length > 0) {
+    ret.push({
+      hotswappable: false,
+      logicalId,
+      rejectedChanges: nonHotswappablePropNames,
+      resourceType: change.newValue.Type,
+    });
```

Comment:
:+1:

The contract is not obvious here, but it's not required because it's autofilled later during printing if unspecified, to reduce code duplication

Reply:
I'd make `classifyChanges` take `ret` too, but it becomes confusing for that method to take a `ChangeHotswapResult`